cache.go

```diff
@@ -0,0 +1,223 @@
+// expands both path to cache and indicator path
+// removes the item if any of path to cache or indicator path is not exist or if the indicator is a dir
+// replaces path to cache (if it is a directory) by every file (recursively) in the directory.
+func normalizeIndicatorByPath(indicatorByPath map[string]string) (map[string]string, error) {
+	normalized := map[string]string{}
+	for pth, indicator := range indicatorByPath {
+		if len(indicator) > 0 {
+			var err error
+			indicator, err = pathutil.AbsPath(indicator)
+			if err != nil {
+				return nil, err
+			}
+
+			info, exist, err := pathutil.PathCheckAndInfos(indicator)
+			switch {
+			case err != nil:
+				return nil, err
+			case !exist:
+				log.Warnf("indicator does not exists at: %s", indicator)
+				continue
+			case info.IsDir():
+				log.Warnf("indicator is a directory: %s", indicator)
+				continue
+			}
+		}
+
+		var err error
+		pth, err = pathutil.AbsPath(pth)
+		if err != nil {
+			return nil, err
+		}
+
+		exist, err := pathutil.IsPathExists(pth)
+		if err != nil {
+			return nil, err
+		}
+		if !exist {
+			log.Warnf("path does not exists at: %s", pth)
+			continue
+		}
+
+		subPths, err := expandPath(pth)
+		if err != nil {
+			return nil, err
+		}
+		for _, p := range subPths {
+			normalized[p] = indicator
+		}
+	}
+	return normalized, nil
+}
+
+// normalizeExcludeByPattern modifies excludeByPattern:
+// expands patterns.
+func normalizeExcludeByPattern(excludeByPattern map[string]bool) (map[string]bool, error) {
+	normalized := map[string]bool{}
+	for pattern, exclude := range excludeByPattern {
+		pattern, err := pathutil.AbsPath(pattern)
+		if err != nil {
+			return nil, err
+		}
+
+		normalized[pattern] = exclude
+	}
+	return normalized, nil
+}
+
+// match reports whether the path matches to any of the given ignore items
+// and returns the exclude property of the matching ignore item.
+func match(pth string, excludeByPattern map[string]bool) (bool, bool) {
+	for pattern, exclude := range excludeByPattern {
+		if strings.Contains(pattern, "*") && glob.Glob(pattern, pth) {
+			return true, exclude
+		}
+
+		if !strings.Contains(pattern, "*") && strings.HasPrefix(pth, pattern) {
+			return true, exclude
+		}
+	}
+	return false, false
+}
+
+// interleave matches the given include items with the ignore items and returns which path needs to be cached:
+// if an ignore item matches to a path, the path either will not affect the previous cache invalidation
+// or will not be included in the cache.
+// Otherwise a path will affect the previous cache invalidation:
+// if the path has indicator, the indicator will affect the previous cache invalidation
+// otherwise the file itself.
+func interleave(indicatorByPth map[string]string, excludeByPattern map[string]bool) (map[string]string, error) {
+	indicatorByCahcePth := map[string]string{}
+
+	for pth, indicator := range indicatorByPth {
+		doNotTrack, exclude := match(pth, excludeByPattern)
```

Comment:
doNotTrack is a strange name

Reply:
what is your suggestion?

Reply:
I meant to impress my feelings about that word, I didn't requested to change it.
If I need to suggest one then I think **skip** would be better.